pkg/agent/proxy/route.go

```diff
@@ -0,0 +1,301 @@
+		exist := false
+		for _, oldItem := range oldRoute {
+			if RouteEqual(oldItem, newItem) {
+				exist = true
+				break
+			}
+		}
+		if !exist {
+			addRoute = append(addRoute, newItem)
+		}
+	}
+
+	// add & del route
+	for i := range delRoute {
+		if err = netlink.RouteDel(&delRoute[i]); err != nil {
+			klog.Errorf("delete route item failed, err: %s", err)
+		} else {
+			klog.Infof("delete route item %s", &delRoute[i])
+		}
+	}
+	for i := range addRoute {
+		if err = netlink.RouteAdd(&addRoute[i]); err != nil {
+			klog.Errorf("add route item failed, err: %s", err)
+		} else {
+			klog.Infof("add route item %s", &addRoute[i])
+		}
+	}
+}
+
+func UpdateIptables(nodeList corev1.NodeList, thisNode corev1.Node) {
+	var exist bool
+	var err error
+
+	ipt, err := iptables.New()
+	if err != nil {
+		klog.Errorf("init iptables error, err: %s", err)
+		return
+	}
+
+	// check existence of chain LYNX-OUTPUT, if not, then create it
+	if exist, err = ipt.ChainExists("nat", "LYNX-OUTPUT"); err != nil {
+		klog.Errorf("Get iptables LYNX-OUTPUT error, error: %s", err)
+		return
+	}
+	if !exist {
+		err = ipt.NewChain("nat", "LYNX-OUTPUT")
+		if err != nil {
+			klog.Errorf("Create iptables LYNX-OUTPUT error, error: %s", err)
+			return
+		}
+	}
+
+	// check and add LYNX-OUTPUT to POSTROUTING
+	if exist, err = ipt.Exists("nat", "POSTROUTING", "-j", "LYNX-OUTPUT"); err != nil {
+		klog.Errorf("Check LYNX-OUTPUT in nat POSTROUTING error, err: %s", err)
+	}
+	if !exist {
+		if err = ipt.Append("nat", "POSTROUTING", "-j", "LYNX-OUTPUT"); err != nil {
+			klog.Errorf("Append LYNX-OUTPUT into nat POSTROUTING error, err: %s", err)
+		}
+	}
+
+	// check and add MASQUERADE in LYNX-OUTPUT"
+	for _, podCIDR := range thisNode.Spec.PodCIDRs {
+		ruleSpec := []string{"-s", podCIDR, "-j", "MASQUERADE"}
+		if exist, err = ipt.Exists("nat", "LYNX-OUTPUT", ruleSpec...); err != nil {
+			klog.Errorf("Check MASQUERADE rule in nat LYNX-OUTPUT error, rule: %s, err: %s", ruleSpec, err)
+			continue
+		}
+		if !exist {
+			err = ipt.Append("nat", "LYNX-OUTPUT", ruleSpec...)
+			if err != nil {
+				klog.Errorf("Add MASQUERADE rule in nat LYNX-OUTPUT error, rule: %s, err: %s", ruleSpec, err)
+				continue
+			}
+		}
+	}
+
+	// check and add ACCEPT in LYNX-OUTPUT"
+	for _, podCIDR := range thisNode.Spec.PodCIDRs {
+		for _, nodeItem := range nodeList.Items {
+			if nodeItem.Name == thisNode.Name {
+				continue
+			}
+			for _, otherPodCIDR := range nodeItem.Spec.PodCIDRs {
+				ruleSpec := []string{"-s", podCIDR, "-d", otherPodCIDR, "-j", "ACCEPT"}
+				if exist, err = ipt.Exists("nat", "LYNX-OUTPUT", ruleSpec...); err != nil {
+					klog.Errorf("Check ACCEPT rule in nat LYNX-OUTPUT error, rule: %s, err: %s", ruleSpec, err)
+					continue
+				}
+				if !exist {
+					if err = ipt.Insert("nat", "LYNX-OUTPUT", 1, ruleSpec...); err != nil {
+						klog.Errorf("Add ACCEPT rule in nat LYNX-OUTPUT error, rule: %s, err: %s", ruleSpec, err)
+						continue
```

Comment:
Anything else need to be handled besides an Error log?

Reply:
These will be checked in each update op. Since the incorrect of iptables will affect the traffic, Is it necessary to be fatal here?

Reply:
Fatal seems not suitable. Is there any other component can monitor the error information and give user a notice?

Reply:
add [Alert] mark for further processing